crates/services/consensus_module/poa/src/sync.rs

```diff
@@ -0,0 +1,377 @@
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum InnerSyncState {
+    /// We are not connected to at least `min_connected_reserved_peers` peers.
+    ///
+    /// InsufficientPeers -> SufficientPeers
+    InsufficientPeers(BlockHeight),
+    /// We are connected to at least `min_connected_reserved_peers` peers.
+    ///
+    /// SufficientPeers -> Synced(...)
+    /// SufficientPeers -> InsufficientPeers(...)
+    SufficientPeers(BlockHeight),
+    /// We can go into this state if we didn't receive any notification
+    /// about new block height from the network for `time_until_synced` timeout.
+    ///
+    /// We can leave this state only in the case, if we received a valid block
+    /// from the network with higher block height.
+    ///
+    /// Synced -> either InsufficientPeers(...) or SufficientPeers(...)
+    Synced {
+        block_height: BlockHeight,
+        has_sufficient_peers: bool,
+    },
+}
+
+impl InnerSyncState {
+    fn from_config(
+        min_connected_reserved_peers: usize,
+        time_until_synced: Duration,
+        block_height: BlockHeight,
+    ) -> Self {
+        match (min_connected_reserved_peers, time_until_synced) {
+            (0, Duration::ZERO) => InnerSyncState::Synced {
+                block_height,
+                has_sufficient_peers: true,
+            },
+            (0, _) => InnerSyncState::SufficientPeers(block_height),
+            _ => InnerSyncState::InsufficientPeers(block_height),
+        }
+    }
+
+    #[cfg(test)]
+    fn block_height(&self) -> &BlockHeight {
+        match self {
+            InnerSyncState::InsufficientPeers(block_height) => block_height,
+            InnerSyncState::SufficientPeers(block_height) => block_height,
+            InnerSyncState::Synced { block_height, .. } => block_height,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{
+        collections::VecDeque,
+        pin::Pin,
+        task::{
+            Context,
+            Poll,
+        },
+        time::Duration,
+    };
+
+    use fuel_core_services::stream::IntoBoxStream;
+
+    struct MockStream<T> {
+        items: VecDeque<T>,
+    }
+
+    impl<T> MockStream<T> {
+        fn new(range: impl IntoIterator<Item = T>) -> Self {
+            Self {
+                items: range.into_iter().collect(),
+            }
+        }
+    }
+
+    impl<T> tokio_stream::Stream for MockStream<T>
+    where
+        T: Unpin,
+    {
+        type Item = T;
+
+        fn poll_next(
+            self: Pin<&mut Self>,
+            _cx: &mut Context<'_>,
+        ) -> Poll<Option<Self::Item>> {
+            let this = self.get_mut();
+            if this.items.is_empty() {
+                Poll::Pending
+            } else {
+                let next_item = this.items.pop_front();
+                Poll::Ready(next_item)
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn test_sync_task() {
```

Comment:
It would be nice to add more tests for different scenarios, but we can do that in the follow-up PR.

Also, we need integration tests to check the behavior of two block producers.